### ISSUE TITLE
Make Backdrop core compatible with PHP 8.2

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5927]
Our Drupal brethren have started some work on this in a meta: https://www.drupal.org/project/drupal/issues/3304807

**Comment by klonos** [comment]
...should we have multiple issues/PRs, or a single PR with multiple commits, and then merge once tests pass?

**Comment by klonos** [comment]
...for those of us using Lando for our locals, php 8.2 is not supported yet. Follow https://github.com/lando/php/issues/40 for updates.

**Comment by argiepiano** [comment]
I think a meta is always best, then developers can tackle each issue separately and update the meta as they do so

**Comment by klonos** [comment]
During this week's dev meeting we have decided to do the following:
- attempt to get a single PR, which (initially only) disables all tests against php 5.6 and 7.4, and switches the php 8 tests to be running against 8.2
- the goal is to have the tests pass, but if the PR gets unwieldy (indicative, above 100 lines of non-trivial changes), then split off to multiple issues/PRs
- once tests pass against php 8.2, we can add back the php 5.6 and 7.4 checks

(point in time during the dev meeting when this issue was discussed: https://youtu.be/vvzsOmfyIsA?t=2624)

**Comment by klonos** [comment]
I have done most of the legwork, which is currently 148 changed files with mostly trivial-to-review changes. Still not ready for review, but there's heaps less tests failing now.

I will start documenting certain things that most likely also apply to us, but unsure how to handle. Starting with this:

https://www.drupal.org/project/drupal/issues/3280033
> PHP has deprecated dollar brace string interpolation in version 8.2 https://wiki.php.net/rfc/deprecate_dollar_brace_string_interpolation

Commit against D9.4: https://git.drupalcode.org/project/drupal/-/commit/6496252

We do have a few such instances of `${`, but not 100% sure how to best handle all of them. I guess that I'll leave that to someone else.

**Comment by klonos** [comment]
...OK, this feels pretty close now, but I'm starting to see red and blue dragons flying around my head 😅 

I have resulted in using `#[\AllowDynamicProperties]` in some instances (mostly for views-related things, which is what the Drupal maintainers of Views 7.x ended up doing), but we can follow up on these to sort them properly later on.

Can someone else now please review my PR so far and take it from here?

Should we merge what we have now, and iterate with other PRs?

Pinging our @backdrop/core-committers

**Comment by klonos** [comment]
...there is a big amount of the following failures:
- `Creation of dynamic property date_sql_handler::$placeholders is deprecated`: not sure how to best address this, but it should knock out a big chunk of failures.
- `Creation of dynamic property [WHATEVER]::$originalLanguage is deprecated`: should we simply add a `protected` for each instance of `$originalLanguage` in each of the failing tests, or is there a "centralized" way to make this go away?
- `Creation of dynamic property Layout::$is_new is deprecated` and `Creation of dynamic property Layout::$original_name is deprecated`: I would like others with deeper knowledge of Layouts to chime in re these. @docwilmot and @argiepiano come to mind, but anyone else with thoughts really. That should also knock out a huge chunk of failures.

**Comment by klonos** [comment]
> ...for those of us using Lando for our locals, php 8.2 is not supported yet. Follow https://github.com/lando/php/issues/40 for updates.

Lando v3.9.0 was just released, which adds support for PHP 8.2 (just tested it on my local). It's a prerelease, so you can't get it with `brew` at the time of writing - you'll need to manually download it from https://github.com/lando/lando/releases

**Comment by indigoxela** [comment]
One question, I can't answer, yet: will we need `#[AllowDynamicProperties]` for entities like Node?

In D7 Node is just a stdClass, and with stdClass dynamic properties are still allowed. However, in B Node (and User, Comment, File, TaxonomyTerm) are fully classed.

But... `abstract class Entity extends stdClass implements EntityInterface` - wait, does that mean that dynamic properties are allowed with core Entities? As they extend stdClass... :thinking:

Why we need that? Take a look at the book module (it uses dynamic properties), same with contrib (poll module, for example). Heck it's even [officially documented that way](https://docs.backdropcms.org/api/backdrop/modules%21examples%21ajax_example%21ajax_example_node_form_alter.inc/1).

Who knows it? Can core entities have dynamic properties, or not?

**Comment by indigoxela** [comment]
@klonos I've left some questions on your PR.

**Comment by herbdool** [comment]
From reading https://wiki.php.net/rfc/deprecate_dynamic_properties I'd say yes. Any class that extends stdClass will allow dynamic properties.

**Comment by argiepiano** [comment]
After some basic testing in PHP 8.2 I can confirm there is no deprecation warning when you use dynamic properties in classes that extend `stdClass`.  Since Backdrop's `Entity` extends `stdClass`, this means that all Backdrop core entities that extend `Entity` (i.e. `User`, `Node`, `Comment`, `File` and `TaxonomyTerm`) will not need any adjustments such as `#[\AllowDynamicProperties]` or others.

**Comment by argiepiano** [comment]
As for the other questions in @klonos [comment above](https://github.com/backdrop/backdrop-issues/issues/5927#issuecomment-1381892546):  I believe, in general there are three options:
- Either make the first parent extend `stdClass`
- or, declare the properties (e.g. it'd be easy to declare `Layout::$is_new` and `Layout::$original_name`)
- or, add `#[\AllowDynamicProperties]` to the top superclass (this may be the way to go with Views handler superclasses like `views_object`)

I think number 1 above may be the simpler way, but I really don't know if there may be any negative repercussions about  changing classes like `Layout` and all views superclasses to extend `stdClass`.

**Comment by herbdool** [comment]
I prefer adding `#[\AllowDynamicProperties]` instead of extending stdClass because it's more obvious and we also don't know the downsides of extending stdClass.

**Comment by indigoxela** [comment]
So, we don't have to do anything with core Entity class and all classes that extend it. That's great! :+1:

Re Simpletest - I think, it's clean (and safe) to declare all properties. A lot of legwork, but no `#[\AllowDynamicProperties]` necessary.

Re Layout - declaring all properties should be possible and fix the problem. Same for other non-views classes in core, I guess.

Re Views: I'd suggest to add `#[\AllowDynamicProperties]` to the base class(es). That reduces the amount of changes and also fixes contrib classes, that extend those base classes. What I read so far, this attribute won't get removed in PHP 9, which gives us time. For PHP 10 we might want to implement __get and __set magic methods in base classes, that still need dynamic properties. Didn't check details for that, yet.

@argiepiano I don't think, switching views base classes to extend stdClass provides more than the new attribute on classes, or the magic methods. Or do you see any extra benefit?

@klonos as very often, you've been ahead of us with starting this issue. Many thanks for your work so far. :pray: I hope, you'll be able to further work on this. :smirk:

**Comment by klonos** [comment]
I am removing myself from the assignee of this issue, since as I also explained in a related thread in Zulip: I want to be helpful, but at the same time I need to be realistic. With the new job, a lot of my energy is being spent on new knowledge and my focus is on that at the moment. Because I am out of my normal routines, I end up being more drained at the end of my work day than before, so I need more time to relax and wind down than usual. I will eventually get back to things, however, if someone else has the energy/capacity, please feel free to either take over from my current PR, or create your own from scratch.

**Comment by indigoxela** [comment]
Good news, it doesn't take much to get Backdrop working without deprecation nagging on PHP 8.2 – for very basic usage, like clicking through pages and admin UI. :grin: Only 9 files need changes for that.

Not so good: of course, the whole testfile legwork is still pending, to add all the properties. But, hey, I expected much worse results.
As far as I can see, currently almost *all* tests fail, but it's just loads and loads of `Creation of dynamic property...` nagging for **test class** properties.

Obviously I started working on this issue. :stuck_out_tongue_winking_eye: 

A question: could it make sense to do the "legwork" in steps? Reviewing hundreds of files at once isn't fun, for sure.

**Comment by argiepiano** [comment]
Personally I'd prefer all these in one PR. The changes are pretty consistent across all files, and typically won't change functionality, so they may be quick to review, even if the changes happen in many files.

**Comment by indigoxela** [comment]
Lots of progress today, but still a lot of work to do. And questions re class TaxonomyVocabulary:

Loads of tests fail because of a dynamic property `TaxonomyVocabulary::$module` – and that's in the upgrade tests.

Could it be, that we need the `$module` property for upgrades from D7?

Also: TaxonomyVocabularyFunctionalTest seems to completely abuse dynamic properties for just the test case. That test probably needs a rewrite. OR... could it be, that class TaxonomyVocabulary might actually need dynamic properties? For which use-cases? That's something I don't feel qualified to answer, so I'd like to get some feedback (help) here, before I dive into a bigger test rewrite.

Many tests pass now, the tricky ones are left (sigh), so if one of you sees something obvious, I might have missed (re class LayoutRelationshipAuthorFromNode or LayoutRendererFlexible or BlockLegacy or Redirect), please let me know.

**Comment by indigoxela** [comment]
Oh, now I get it, the weird stuff with vocabx happens in function taxonomy_form_vocabulary_submit(). :scream_cat:

@hosef do you happen to remember, what the reason for this permissive loop was? https://github.com/backdrop/backdrop/blob/1.x/core/modules/taxonomy/taxonomy.admin.inc#L359 At this point $form_state['values'] contains all existing roles (??), for example.

**Comment by argiepiano** [comment]
> @hosef do you happen to remember, what the reason for this permissive loop was?

This is my impression of this permissive loop: originally this property assignment was done with  `entity_form_submit_build_entity()`, which includes such a loop. The problem is that vocabularies are not entities in Backdrop, so this probably caused some issues (e.g. running `entity_get_info($entity_type);` for vocabularies).  So, the loop was extracted from that function and put into the submit handler for vocabulary editing/creation. 

And I believe these types of loops were done willy-nilly when it was ok to have dynamic properties that would never be declared or even used (when you save. So, it was ok to throw "everything and the kitchen sink" into the `$vocabulary` object, including roles, etc etc. I think we have to be much more selective now that we have to be careful about using pre-declared properties. 

I would suggest to replace that loop with very specific property assignments. You can get those from `TaxonomyVocabulary::save()`. The only properties that should be assigned in the submit handler are: name, machine_name, description, hierarchy, language, weight.

Alternatively (I just saw your edit), it'd be a good idea to check if the property exists before assigning it in the loop.

**Comment by indigoxela** [comment]
> ... it was ok to throw "everything and the kitchen sink" into the $vocabulary...

@argiepiano hehe, that was my conclusion, too. :grin:  Doing [a check](https://github.com/backdrop/backdrop/pull/4445/commits/745d18b2bc520ee49a98f8d54b7782d658443c12) fixes that.

So we're probably down now to the really complicated cases. :stuck_out_tongue_winking_eye: And I definitely need a pause now. :wave:

**Comment by argiepiano** [comment]
> Loads of tests fail because of a dynamic property TaxonomyVocabulary::$module – and that's in the upgrade tests.

Can you clarify which tests use `module` as a property of vocabularies? I'm not seeing any instance of `module` in the tests in `taxonomy.test`.

**Comment by argiepiano** [comment]
I think the best way to fix the `module` property assignment issue upon upgrades is to modify `taxonomy_update_1001()` to unset `$vocabulary_data['module']`. Currently it only unsets `vid`. Since `module` is not used in Backdrop at all, that should also be unset.

**Comment by klonos** [comment]
Wow @indigoxela! ...great work! 👍🏼  🙏🏼 ...this is in a so much better place now. Thank you for working on it   🙏🏼  🙏🏼 ❤️

**Comment by indigoxela** [comment]
> I think the best way to fix the module property assignment issue upon upgrades is to modify taxonomy_update_1001()

@argiepiano many thanks for your hint, this is really helpful! :pray:

**Comment by indigoxela** [comment]
All tests are green now, ready for testing and review! :tada:

**Comment by klonos** [comment]
Awesome! ...I did a quick pass and left some questions in the PR. Please excuse ignorance if these are obvious things to others.

**Comment by indigoxela** [comment]
> I did a quick pass and left some questions in the PR.

Many thanks for having a look! :pray: I hope, my answers are helpful. No need to excuse for normal questions, BTW. :wink:

I'll add the requested inline code comment later.

**Comment by klonos** [comment]
Thanks @indigoxela ...I've added some further comments/replies, but these don't necessarily need any action from you now. Let's wait for others to have a look and see if there's anything else needed.

It does seem to me that since we are fixing things, we might as well fix them consistently, but I understand your concerns re the size of the PR and the amount of files/lines changed.

**Comment by indigoxela** [comment]
> It does seem to me that since we are fixing things, we might as well fix them consistently

@klonos  I understand your desire for consistency. And I'm glad, it's OK for you to not start this discussion about BackdropDateTime in a PR, which is already quite big, although it fixes only one task. :pray:

**Comment by indigoxela** [comment]
> I couldn't review all of it, so I only added questions for a tiny portion...

@argiepiano That's understood. Reviewing the whole change is a bigger task, better done in steps. Many thanks for your hints so far. :+1:

**Comment by argiepiano** [comment]
@indigoxela, thanks for all this work. I'm wondering what are your thoughts about `view::date_info`? I didn't see that addressed in the PR.  A couple of date handlers (pager and arguments) and theming functions in the date module use this property to store information in the view object for further retrieval down the line. I've seen this approach also for some custom module views handlers (I believe in Ubercart and others). 

I think it would not be efficient to create a `date_info` property in the top `view` class, as this is only needed by date. However, I'm thinking it may be a good idea to create a "catch all" property there to store stuff as arrays, e.g. `view::infoStorage`, and then make changes to date and other modules that use dynamic properties in the top `view` class so that they store their stuff within that property, as in `$view->infoStorate['date_info']` etc. What do you think?

**Comment by herbdool** [comment]
Is `view::date_info` a public or protected property? I don't think we should be changing it if so.

**Comment by argiepiano** [comment]
> Is view::date_info a public or protected property? I don't think we should be changing it if so.

It's never declared. It's public by default. The issue here is that it's used only by the date module to attach information to the `view` object that only date uses, so adding it as a public property of `view` class doesn't make a lot of sense. 

This approach (adding dynamic properties to `view` as a temporary storage [NB: these are not saved]) is used by a few contrib modules as well. My suggestion was to create a public property of `view` that could be used by `date` and other modules to attach temporary information to the object.

**Comment by argiepiano** [comment]
@indigoxela, never mind my question above. I continued reading the PR and noticed you added `#[AllowDynamicProperties]` to class `view`, so that will take care of this and other dynamic properties.

**Comment by herbdool** [comment]
Oh right, it isn't declared. Oops,  weekend brain. Sounds like it's settled anyway.

**Comment by indigoxela** [comment]
@argiepiano yes, several plugins use dynamic properties in views. That's why our solution is, to allow that via `#[AllowDynamicProperties]`.

Adding the date_info to all views, doesn't seem efficient to me, because - as you already stated - only the date module and possibly some contribs make use of it. Adding a generic `$view->infoStorage` sounds like a promising approach to get rid of dynamic properties in $views, but... that's a bigger change and a trickier re backwards compatibility. And probably out of scope for this issue here. We try not to break anything for contrib and custom. :wink: It might eventually even go into a bugfix release, as far as I can see.

My personal suggestion would be a follow-up, to discuss this properly. There might be some pitfalls.

**Comment by klonos** [comment]
I would like us to merge this into 1.x as soon as possible. Lets iterate and improve please, rather than try to get everything perfect in a single go. We can declare php 8.2 compatibility in a bug fix release, and then call any follow-up either a task or a bug and keep improving/fixing things.

I know that @herbdool has 👀 here, but pinging our other @backdrop/core-committers to please have a look as well.

**Comment by klonos** [comment]
...in case it wasn't clear from my previous comment, my main goal is to merge this into 1.x - not necessarily to make a release for it now. Just looking to get more exposure and testing via our PR sandboxes.

**Comment by indigoxela** [comment]
Hm... just saw that there's a conflict re the .tugboat/config.yml, so I resolved it. And the tugboat sandbox refused to get created. With exactly the same value which has worked before. :roll_eyes: 

It could eventually have to do with the fact that Debian creates a new release today/now. Not sure, where Tugboat pull their stuff from.
Anyway – we don't have a PR sandbox. :shrug:

**Comment by yorkshire-pudding** [comment]
I've tested this locally with a few modules; created webform and exported config; created paragraph types, added paragraphs (found a bug in one of my modules)
All works well.
I haven't tested `bee` as had to create a custom lando instance rather than using my usual script, but that will be easier to test once this is merged.

**Comment by klonos** [comment]
@indigoxela I've left a comment in the PR re using the `8.2-apache` image instead of `8.2-apache-bullseye` (or any other OS-specific image). @quicksketch has requested that we do that when we merged #6133.

**Comment by indigoxela** [comment]
@klonos I removed the OS part as requested and verified, that we still have php 8.2. :+1:

**Comment by indigoxela** [comment]
Friendly reminder that this PR still needs a final review (a lot has already happened a month ago). :raised_hands: 

If someone's concerned because of the failing phpcs check: that's just because that GHA can't handle such big changes (82 files). I'm pretty sure, all my changes comply with our coding standards. :grin:

**Comment by herbdool** [comment]
A couple more tweaks and I think it's ready @indigoxela. Thanks for your work on this.

**Comment by indigoxela** [comment]
@herbdool that's great! Many thanks for reviewing, I've addressed your requests.

**Comment by herbdool** [comment]
I think it's good to go!

**Comment by quicksketch** [comment]
I code reviewed this and it looks great! Nice job @indigoxela defining all those missing properties! This will really substantially reduce the amount of warnings I get in PHPStorm, and help with type-hinting as well. All around a big set of improvements and clean-up!! This is a big effort and thank you so much for the work.

The only thing I found in my read-through was that we're not consistent with `#[AllowDynamicProperties]` vs `#[\AllowDynamicProperties]`. More in the code review here: https://github.com/backdrop/backdrop/pull/4445#pullrequestreview-1529349216

**Comment by quicksketch** [comment]
I added a follow-up issue regarding whether we should use a backslash or not when defining class attributes here: https://github.com/backdrop/backdrop-issues/issues/6169

**Comment by indigoxela** [comment]
> The only thing I found in my read-through was that we're not consistent with #[AllowDynamicProperties]

Right, that's an oversight and fixed now, to be consistent. Many thanks for pointing me to that and for reviewing. :pray:

**Comment by quicksketch** [comment]
Looks like @herbdool also re-approved this in the PR. Looks good to me! I'll let tests finish running to be safe.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4445 into 1.x and 1.25.x (no conflicts!). Thank you so much @indigoxela, @herbdool, @argiepiano, @klonos, and @yorkshire-pudding for the extensive work put into this! I really like where this ended and the updates will have a lot of other peripheral benefits, like better code hinting and IDE support. Very nice!
